src/legacy/core_plugins/embeddable_api/public/containers/container.ts

```diff
@@ -0,0 +1,362 @@
+    });
+
+    return ({
+      ...containerInput,
+      ...explicitFiltered,
+      // Typescript has difficulties with inferring this type but it is accurate with all
+      // tests I tried. Could probably be revisted with future releases of TS to see if
+      // it can accurately infer the type.
+    } as unknown) as TEmbeddableInput;
+  }
+
+  public destroy() {
+    super.destroy();
+    Object.values(this.children).forEach(child => child.destroy());
+    this.subscription.unsubscribe();
+  }
+
+  protected createNewPanelState<
+    TEmbeddableInput extends EmbeddableInput,
+    TEmbeddable extends IEmbeddable<TEmbeddableInput, any>
+  >(
+    factory: EmbeddableFactory<TEmbeddableInput, any, TEmbeddable>,
+    partial: Partial<TEmbeddableInput> = {}
+  ): PanelState<TEmbeddableInput> {
+    const embeddableId = partial.id || uuid.v4();
+
+    const explicitInput = this.createNewExplicitEmbeddableInput<TEmbeddableInput>(
+      embeddableId,
+      factory,
+      partial
+    );
+
+    return {
+      type: factory.type,
+      embeddableId,
+      explicitInput: {
+        id: embeddableId,
+        ...explicitInput,
+      },
+    };
+  }
+
+  protected getPanelState<TEmbeddableInput extends EmbeddableInput = EmbeddableInput>(
+    embeddableId: string
+  ) {
+    if (this.input.panels[embeddableId] === undefined) {
+      throw new Error(`No embeddable with id ${embeddableId}, this  ${JSON.stringify(this.input)}`);
+    }
+    const panelState: PanelState = this.input.panels[embeddableId];
+    return panelState as PanelState<TEmbeddableInput>;
+  }
+
+  protected abstract getInheritedInput(id: string): TChildInput;
+
+  private createNewExplicitEmbeddableInput<
+    TEmbeddableInput extends EmbeddableInput = EmbeddableInput,
+    TEmbeddable extends IEmbeddable<TEmbeddableInput, EmbeddableOutput> = IEmbeddable<
+      TEmbeddableInput
+    >
+  >(
+    id: string,
+    factory: EmbeddableFactory<TEmbeddableInput, any, TEmbeddable>,
+    partial: Partial<TEmbeddableInput> = {}
+  ): Partial<TEmbeddableInput> {
+    const inheritedInput = this.getInheritedInput(id);
+    const defaults = factory.getDefaultInput(partial);
+
+    // Container input overrides defaults.
+    const explicitInput: Partial<TEmbeddableInput> = partial;
+
+    getKeys(defaults).forEach(key => {
+      // @ts-ignore We know this key might not exist on inheritedInput.
+      const inheritedValue = inheritedInput[key];
+      if (inheritedValue === undefined && explicitInput[key] === undefined) {
+        explicitInput[key] = defaults[key];
+      }
+    });
+    return explicitInput;
+  }
+
+  private onPanelRemoved(id: string) {
+    // Clean up
+    const embeddable = this.getChild(id);
+    if (embeddable) {
+      embeddable.destroy();
+
+      // Remove references.
+      delete this.children[id];
+    }
+
+    // Update output state.
+    this.updateOutput({
+      embeddableLoaded: {
+        ...this.output.embeddableLoaded,
+        [id]: undefined,
```

Comment:
If you want to remove `[id]` maybe use the "omit" trick. Setting the key to `undefined` does not remove it.

All these state manipulations are akin to Redux reducers.